src/Protobuf/Encode.elm

```diff
@@ -565,40 +622,36 @@ tag fieldNumber wireType =
 -- VARINT
 
 
-varInt : Int -> ( Int, Encode.Encoder )
-varInt value =
-    let
-        encoders =
-            toVarIntEncoders value
-    in
-    ( List.length encoders, Encode.sequence encoders )
+intEncoder : EncodeVarInt intType config -> intType -> Encoder
+intEncoder config =
+    Encoder VarInt << varInt config
 
 
-unsigned : Int -> Int
-unsigned value =
-    if value >= 2 ^ 31 then
-        value - 2 ^ 32
+sintEncoder : EncodeVarInt intType { config | zigZag : intType -> intType } -> intType -> Encoder
+sintEncoder config =
+    intEncoder config << config.zigZag
 
-    else
-        value
 
+uintEncoder : EncodeVarInt intType { config | fromUnsigned : intType -> intType } -> intType -> Encoder
+uintEncoder config =
+    intEncoder config << config.fromUnsigned
 
-zigZag : Int -> Int
-zigZag value =
-    Bitwise.xor (Bitwise.shiftRightBy 31 value) (Bitwise.shiftLeftBy 1 value)
 
+varInt32 : Int -> ( Int, Encode.Encoder )
+varInt32 =
+    varInt Int32.config
 
-toVarIntEncoders : Int -> List Encode.Encoder
-toVarIntEncoders value =
-    let
-        base128 =
-            Bitwise.and 0x7F value
 
-        higherBits =
-            Bitwise.shiftRightZfBy 7 value
+varInt : EncodeVarInt intType config -> intType -> ( Int, Encode.Encoder )
+varInt config value =
+    let
+        encoders =
+            config.to7BitList value |> List.map Encode.unsignedInt8
     in
-    if higherBits /= 0x00 then
-        Encode.unsignedInt8 (Bitwise.or 0x80 base128) :: toVarIntEncoders higherBits
+    ( List.length encoders, Encode.sequence encoders )
+
 
-    else
-        [ Encode.unsignedInt8 base128 ]
+type alias EncodeVarInt intType config =
```

Comment:
Why have you introduced this `EncodeVarInt`? And why do all the functions use this one rather than the internal `IntType` that you have defined?

Reply:
I guess I can get rid of these now. That was my attempt to achieve 2 goals
1. a interface for Int and Int64 (IntType)
2. not forcing the user to implement all of the interfaces methods if they just want to decode, say, unsigned ints into Bytes.

Since 2. is a non-issue now, after we hid the exposed `intDecoder` functions, I can remove the sub-interfaces.